src/batch.c

```diff
@@ -0,0 +1,1154 @@
+		bool active;
+		bool scripts;
+	} Q;
+	struct BATCH_S {	/* -Sb|f<script> */
+		bool active;
+		char *file;	/* Name of script file */
+		FILE *fp;	/* Open file pointer to script */
+	} S[2];
+	struct BATCH_T {	/* -T<n_jobs>|<min>/<max/<inc>[+n]|<timefile>[+p<precision>][+s<job>][+w] */
+		bool active;
+		bool split;		/* true means we must split any trailing text in to words */
+		unsigned int n_jobs;	/* Total number of jobs */
+		unsigned int start_job;	/* First job [0] */
+		unsigned int precision;	/* Decimals used in making unique job tags */
+		char *file;		/* timefile name */
+	} T;
+	struct BATCH_W {	/* -W<workingdirectory> */
+		bool active;
+		char *dir;	/* Alternative working directory than implied by -N */
+	} W;
+	struct BATCH_Z {	/* -Z[ */
+		bool active;	//* Delete all files including the mainscript and anything passed via -I, -S */
+	} Z;
+	struct BATCH_x {	/* -x[[-]<ncores>] */
+		bool active;
+		int n_threads;
+	} x;
+};
+
+struct BATCH_STATUS {
+	/* Used to monitor the start, running, and completion of job jobs running in parallel */
+	bool started;	/* true if job job has started */
+	bool completed;	/* true if the completion file has been successfully produced */
+};
+
+static void *New_Ctrl (struct GMT_CTRL *GMT) {	/* Allocate and initialize a new control structure */
+	struct BATCH_CTRL *C;
+
+	C = gmt_M_memory (GMT, NULL, 1, struct BATCH_CTRL);
+	C->x.n_threads = GMT->parent->n_cores;	/* Use all cores available unless -x is set */
+	return (C);
+}
+
+static void Free_Ctrl (struct GMT_CTRL *GMT, struct BATCH_CTRL *C) {	/* Deallocate control structure */
+	gmt_M_unused (GMT);
+	if (!C) return;
+	gmt_M_str_free (C->In.file);
+	gmt_M_str_free (C->I.file);
+	gmt_M_str_free (C->N.prefix);
+	gmt_M_str_free (C->S[BATCH_PREFLIGHT].file);
+	gmt_M_str_free (C->S[BATCH_POSTFLIGHT].file);
+	gmt_M_str_free (C->T.file);
+	gmt_M_str_free (C->W.dir);
+	gmt_M_free (GMT, C);
+}
+
+/*! -x[[-]<ncores>] parsing needed but here not related to OpenMP etc - it is just a local option */
+GMT_LOCAL int batch_parse_x_option (struct GMT_CTRL *GMT, struct BATCH_CTRL *Ctrl, char *arg) {
+	if (!arg) return (GMT_PARSE_ERROR);	/* -x requires a non-NULL argument */
+	if (arg[0])
+		Ctrl->x.n_threads = atoi (arg);
+
+	if (Ctrl->x.n_threads == 0)	/* Not having any of that.  At least one */
+		Ctrl->x.n_threads = 1;
+	else if (Ctrl->x.n_threads < 0)	/* Meant to reduce the number of threads */
+		Ctrl->x.n_threads = MAX(GMT->parent->n_cores + Ctrl->x.n_threads, 1);		/* Max-n but at least one */
+	return (GMT_NOERROR);
+}
+
+GMT_LOCAL int batch_gmt_sleep (unsigned int microsec) {
+	/* Waiting before checking if the completion file has been generated */
+#ifdef WIN32
+	Sleep ((uint32_t)microsec/1000);	/* msec are microseconds but Sleep wants milliseconds */
+	return 0;
+#else
+	return (usleep ((useconds_t)microsec));
+#endif
+}
+
+GMT_LOCAL void batch_set_value (struct GMT_CTRL *GMT, FILE *fp, int mode, int col, char *name, double value) {
```

Comment:
These functions are the same as those in movie.c, right? Can we reuse these functions?

Reply:
Yes, I thought about that as I was doing my copy/cut.  Will check.

Reply:
In trying not to duplicate code (always a good thing), here are some options, given that I would like to follow our static function naming convention as well.  Let's look at the functions called

```
GMT_LOCAL int batch_gmt_sleep (unsigned int microsec)
GMT_LOCAL int movie_gmt_sleep (unsigned int microsec)

```
I would like one code base yielding two separate static functions with the above names.  I think these are viable options:

1. Make a gmt_script.c.in (or h.in?) which lists `GMT_LOCAL int @MODULE@_gmt_sleep (unsigned int microsec)` etc, and let CMake build both batch_sub.c and movie_sub.c from that.
2. Do some trickery with C preprocessors.  If I remember right, the #define statements is kind of like text replacement.  If so, might we set #define THIS movie in movie.c and have gmt_script.c be included and it will list G`MT_LOCAL int THIS_gmt_sleep (unsigned int microsec)`.  Will that work?
3. Just elevate these to gmtlib_* functions and place them in gmt_support.c and they can be used by both batch.c, movie.c, and possibly other future modules.

Perhaps there is no real savings and possibly confusion by **not** doing optino 3?

Reply:
Looks like the price to pay for keeping them static is too much, @seisman ?

Reply:
Option 3 seems the easiest one to maintain.

Reply:
Great. Going with that.

Reply:
Done.